### PR TITLE
Implement and Verify Proposal Signatures

### PR DESCRIPTION
**Pull request by IvanTheGreatDev**
Part of #1366 to add block signature validation to the client, proposal signatures were previously stubs and this now allows block signature creation and verification.

### CODE REVIEW

File: beacon-chain/blockchain/service_test.go [hunk 1]
```diff
@@ -165,7 +165,7 @@ func createPreChainStartDeposit(t *testing.T, pk []byte) *pb.Deposit {
 
 func createRandaoReveal(t *testing.T, beaconState *pb.BeaconState, privKeys []*bls.SecretKey) []byte {
 	// We fetch the proposer's index as that is whom the RANDAO will be verified against.
-	proposerIdx, err := helpers.BeaconProposerIndex(beaconState, beaconState.Slot)
+	proposerIndex, err := helpers.BeaconProposerIndex(beaconState, beaconState.Slot)
```
[the remark this answers is not in the document]
Comment (IvanTheGreatDev) [reply beacon-chain/blockchain/service_test.go]: Mainly to be consistent with all the signing functions

File: beacon-chain/blockchain/service_test.go [hunk 2]
```diff
@@ -439,6 +439,16 @@ func TestChainService_Starts(t *testing.T) {
 		},
 	}
 
+	proposerIndex, err := helpers.BeaconProposerIndex(beaconState, block.Slot)
+	if err != nil {
+		t.Errorf("could not get beacon proposer index: %v", err)
```
[the remark this answers is not in the document]
Comment (IvanTheGreatDev) [reply beacon-chain/blockchain/service_test.go]: The slot isn't logged out for the other instances of this error (like in randao sigs), should I change those as well?

[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/blockchain/service_test.go]: yeah that would be better than a generic error message

File: beacon-chain/blockchain/service_test.go [hunk 3]
```diff
@@ -519,10 +529,22 @@ func TestReceiveBlock_RemovesPendingDeposits(t *testing.T) {
 	}
 
 	beaconState.Slot--
+
+	proposerIndex, err := helpers.BeaconProposerIndex(beaconState, block.Slot)
+	if err != nil {
+		t.Errorf("could not get beacon proposer index at slot %v: %v", block.Slot, err)
```
Comment (terenc3t) [inline beacon-chain/blockchain/service_test.go]:
```suggestion
		t.Errorf("could not get beacon proposer index at slot %d: %v", block.Slot, err)
```

File: beacon-chain/core/blocks/block.go [hunk 4]
```diff
@@ -103,3 +106,24 @@ func BlockChildren(block *pb.BeaconBlock, observedBlocks []*pb.BeaconBlock) ([]*
 	}
 	return children, nil
 }
+
+// BlockSignature returns the signature of the block using the secret key provided
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block.go]:
```suggestion
// BlockSignature returns the signature of the block using the secret key provided.
```

File: beacon-chain/core/blocks/block.go [hunk 5]
```diff
@@ -103,3 +106,24 @@ func BlockChildren(block *pb.BeaconBlock, observedBlocks []*pb.BeaconBlock) ([]*
...
+		return []byte{}, fmt.Errorf("could not hash beacon block: %v", err)
+	}
+
+	proposalHash, err := hashutil.HashProposal(&pb.Proposal{
+		Slot:            block.Slot,
+		Shard:           params.BeaconConfig().BeaconChainShardNumber,
+		BlockRootHash32: blockRootHash[:],
+	})
+	if err != nil {
+		return []byte{}, fmt.Errorf("could not hash attestation data: %v", err)
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block.go]: this is not `attestation data` is it?

File: beacon-chain/core/blocks/block_operations.go [hunk 6]
```diff
@@ -31,13 +31,44 @@ import (
...
+	if err != nil {
+		return fmt.Errorf("could not hash attestation data: %v", err)
+	}
+
+	sig, err := bls.SignatureFromBytes(block.Signature)
+	currentEpoch := helpers.CurrentEpoch(beaconState)
+
+	domain := forkutils.DomainVersion(beaconState.Fork, currentEpoch, params.BeaconConfig().DomainProposal)
+
+	log.WithFields(logrus.Fields{}).Info("Verifying proposal")
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block_operations.go]: Please don't `Info` this. Let's use `debug` and add more context to the log `Verifying proposal hash %#x etc`

File: beacon-chain/core/state/transition_test.go [hunk 7]
```diff
@@ -42,16 +43,16 @@ func setupInitialDeposits(t *testing.T, numDeposits uint64) ([]*pb.Deposit, []*b
 
 func createRandaoReveal(t *testing.T, beaconState *pb.BeaconState, privKeys []*bls.SecretKey) []byte {
 	// We fetch the proposer's index as that is whom the RANDAO will be verified against.
-	proposerIdx, err := helpers.BeaconProposerIndex(beaconState, beaconState.Slot)
+	proposerIndex, err := helpers.BeaconProposerIndex(beaconState, beaconState.Slot)
 	if err != nil {
-		t.Fatal(err)
+		t.Errorf("could not get beacon proposer index at slot %v: %v", beaconState.Slot, err)
```
Comment (terenc3t) [inline beacon-chain/core/state/transition_test.go]:
```suggestion
		t.Errorf("could not get beacon proposer index at slot %d: %v", beaconState.Slot, err)
```

File: beacon-chain/core/state/transition_test.go [hunk 8]
```diff
@@ -232,12 +192,20 @@ func TestProcessBlock_IncorrectAggregateSig(t *testing.T) {
...
-			AttesterSlashings: attesterSlashings,
-			Attestations:      attestations,
-			VoluntaryExits:    exits,
+			Attestations: attestations,
 		},
 	}
+
+	proposerIndex, err := helpers.BeaconProposerIndex(beaconState, block.Slot)
+	if err != nil {
+		t.Errorf("could not get beacon proposer index at slot %v: %v", block.Slot, err)
```
Comment (terenc3t) [inline beacon-chain/core/state/transition_test.go]:
```suggestion
		t.Errorf("could not get beacon proposer index at slot %d: %v", block.Slot, err)
```

File: beacon-chain/core/state/transition_test.go [hunk 9]
```diff
@@ -498,11 +466,50 @@ func TestProcessBlock_PassesProcessingConditions(t *testing.T) {
 			VoluntaryExits:    exits,
 		},
 	}
+
+	proposerIndex, err := helpers.BeaconProposerIndex(beaconState, block.Slot)
+	if err != nil {
+		t.Errorf("could not get beacon proposer index at slot %v: %v", block.Slot, err)
```
Comment (terenc3t) [inline beacon-chain/core/state/transition_test.go]:
```suggestion
		t.Errorf("could not get beacon proposer index at slot %d: %v", block.Slot, err)
```

File: beacon-chain/core/state/transition_test.go [hunk 10]
```diff
@@ -498,11 +466,50 @@ func TestProcessBlock_PassesProcessingConditions(t *testing.T) {
...
+	}
+
+	beaconState.Slot = params.BeaconConfig().GenesisSlot + 10
+	block := &pb.BeaconBlock{
+		Slot: params.BeaconConfig().GenesisSlot + 10,
+	}
+
+	proposerIndex, err := helpers.BeaconProposerIndex(beaconState, block.Slot)
+	if err != nil {
+		t.Errorf("could not get beacon proposer index at slot %v: %v", block.Slot, err)
```
Comment (terenc3t) [inline beacon-chain/core/state/transition_test.go]:
```suggestion
		t.Errorf("could not get beacon proposer index at slot %d: %v", block.Slot, err)
```

File: validator/client/validator_propose.go [hunk 11]
```diff
@@ -115,8 +115,27 @@ func (v *validator) ProposeBlock(ctx context.Context, slot uint64) {
 	block.StateRootHash32 = resp.GetStateRoot()
 
 	// 4. Sign the complete block.
-	// TODO(1366): BLS sign block
-	block.Signature = nil
+	blockRootHash, err := hashutil.HashBeaconBlock(block)
+	if err != nil {
+		log.Errorf("Could not hash beacon block: %v", err)
```
Comment (terenc3t) [inline validator/client/validator_propose.go]:
```suggestion
		log.Errorf("Could not hash beacon block slot %d: %v",block.slot, err)
```

File: validator/client/validator_propose.go [hunk 12]
```diff
@@ -115,8 +115,27 @@ func (v *validator) ProposeBlock(ctx context.Context, slot uint64) {
...
+		Slot:            slot,
+		Shard:           params.BeaconConfig().BeaconChainShardNumber,
+		BlockRootHash32: blockRootHash[:],
+	})
+	if err != nil {
+		log.Errorf("Could not hash proposal signed data: %v", err)
+		return
+	}
+
+	log.Infof("Signing proposal for slot: %d", slot)
```
Comment (terenc3t) [inline validator/client/validator_propose.go]: Use `Debug`

File: validator/client/validator_propose.go [hunk 13]
```diff
@@ -115,8 +115,27 @@ func (v *validator) ProposeBlock(ctx context.Context, slot uint64) {
...
+	proposalDomain := forkutils.DomainVersion(fork, epoch, params.BeaconConfig().DomainProposal)
+	proposalSignature := v.key.SecretKey.Sign(proposalDataHash[:], proposalDomain)
+	block.Signature = proposalSignature.Marshal()
+	log.Infof("Proposal signature: %#x", proposalSignature.Marshal())
```
Comment (terenc3t) [inline validator/client/validator_propose.go]: Use `Debug`

**Pull request closed without merging**